e2e/LongHaul/service/IotHub.cs

```diff
@@ -124,78 +71,37 @@ public async Task MonitorConnectedDevicesAsync(CancellationToken ct)
 
                 await foreach (ClientTwin device in allDevices)
                 {
-                    if (s_activeDeviceSet.Contains(device.DeviceId) && device.ConnectionState == ClientConnectionState.Disconnected)
+                    string deviceId = device.DeviceId;
+
+                    if (s_onlineDeviceOperations.ContainsKey(deviceId) && device.ConnectionState is ClientConnectionState.Disconnected)
                     {
-                        s_activeDeviceSet.Remove(device.DeviceId);
+                        s_onlineDeviceOperations.Remove(deviceId);
                     }
-                    else if (!s_activeDeviceSet.Contains(device.DeviceId) && device.ConnectionState == ClientConnectionState.Connected)
+                    else if (!s_onlineDeviceOperations.ContainsKey(deviceId) && device.ConnectionState is ClientConnectionState.Connected)
                     {
-                        s_activeDeviceSet.Add(device.DeviceId);
+                        s_onlineDeviceOperations.Add(
+                            deviceId,
+                            async () =>
+                            {
+                                using var deviceOperations = new DeviceOperations(s_serviceClient, deviceId, _logger);
```

Comment:
Clone the logger, add the device Id as a custom dimension, and pass that instance to this class.
```suggestion
                                using var deviceOperations = new DeviceOperations(s_serviceClient, deviceId, _logger);
```